content/posts/pandastutor.md

```diff
@@ -0,0 +1,323 @@
+
+
+## What is Pandas Tutor?
+
+[Pandas](https://pandas.pydata.org/) is now an industry-standard tool
+for data science, but as teachers we've seen firsthand how it can be
+hard for students to learn due to its [complex
+API](https://pandas.pydata.org/docs/reference/index.html). For example,
+this code selects, sorts, and groups values from a `dogs` dataset to
+produce a summary table. But if you run this code in a Jupyter notebook,
+all you see is the end result:
+
+<img src="pandastutor-dogs.png" style="width: 70%;"/>
+
+Wait, what exactly did pandas do to turn the `dogs` dataset into this
+output summary table? It's not at all clear from what you see above.
+
+If you ran this exact same code in our [Pandas
+Tutor](https://pandastutor.com/) tool, it will show you exactly what's
+going on step-by-step as your code selects, sorts, groups, and
+calculates the group medians:
+
+<img src="pandastutor-intro-animated.gif"/>
+
+[Try this example live
+here](https://pandastutor.com/vis.html#trace=example-code/py_dogs.json).
+
+Instructors can use Pandas Tutor to aid their teaching, and students can
+use it to understand and debug their homework assignments.
+
+
+## Why did we port Pandas Tutor to Pyodide?
+
+The original version of Pandas Tutor ran the user's code on our Linux
+server by spinning up a new Docker container for every request. It main
+limitation was slowness -- from the time a user hits the 'Run' button,
+it takes up to 5 seconds for the server to run their code, produce a
+step-by-step execution trace, and send it over the network to their
+browser. And when there's a lot of concurrent users the latency may be
+up to 10 seconds, or the server might crash due to too many Docker
+containers. This can be a frustrating user experience when an instructor
+is lecturing or when hundreds of students are using the tool at the same
+time to visualize in-class code examples.
+
+Beyond user experience, this server-based setup was also a pain for us
+as system maintainers since we had to carefully configure our Linux and
+Docker setups to handle ever-increasing scale. Remember, we're
+university instructors, **not** professional software developers or
+DevOps staff! That means we don't have the know-how or institutional
+resources required to maintain a production-scale server deployment
+setup.
+
+That's why we were thrilled when we discovered
+[Pyodide](https://pyodide.org/) since it could help us overcome all
+these limitations. Specifically, now that we ported Pandas Tutor over to
+use Pyodide:
+
+- Users see their visualizations **instantly** since all their code runs
+  in-browser. No more waiting 5-10 seconds for server-side execution and
+  network data transfer. A major source of server slowdown was starting
+  up a new Python interpreter and loading all dependencies like pandas
+  and numpy *every time a user runs a piece of code*. With Pyodide this
+  startup happens only once per browser session, not every execution.
```

Comment:
This  point might need to be updated a bit. There is still a startup cost for Pyodide, particularly when loading pandas which I guess should be around 15 s. Then once it's loaded the visualization is indeed fast. 

Also, I understand that the previous setup was much slower, but I'm not sure this could be fully attributed to the interpreter startup,
```
time python -c "import pandas"

real    0m0,220s
user    0m0,345s
sys     0m0,522s
```
on my laptop with SSD, so nowhere near 5s. Also if the code itself runs instantly on pyodide, it would also run instantly on the server since Pyodide is slower than native runtime. So that leaves serializing and network transfer. That adds some overhead, but still 5-10 s feels like a lot.

So my point is running it server-side could theoretically be done with sub second latency given enough engineering resources, but it's also not what you want/can spend your time on (as you say below).

So I think it's good to have this bullet point, but the arguments/wording used might need to be adjusted a bit.

Reply:
yep, good point, will shorten and adjust. our server setup is fairly sub-optimal due to, as you mentioned, the implementation effort of streamlining it (while keeping things nicely sandboxed, etc.)